### PR TITLE
Custom User Settings

### PR DESCRIPTION
**Pull request by deanmarcussen**
I don't think we have an issue about this, but we've talked about it a few times.

Custom User Settings, much like Custom Site Settings.

Implement a content type with a stereotype `CustomUserSettings` and it'll show up on a tab in the user properties.

Screenshot 2020-11-09 at 14 21 54

Screenshot 2020-11-09 at 14 23 10

Screenshot 2020-11-09 at 14 23 16

I've just done this as a feature of `Users` as it's a small number of files required.

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Contents/Views/AdminTabs.cshtml [hunk 1]
```diff
@@ -9,7 +9,7 @@
     <ul class="nav nav-tabs flex-column flex-md-row" role="tablist">
         @foreach (var tab in tabs)
         {
-            var tabId = $"tab-{tab}-{Model.ContentItem.ContentItemId}".HtmlClassify();
+            var tabId = $"tab-{tab}-{Model.ContentItem?.ContentItemId}".HtmlClassify();
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Contents/Views/AdminTabs.cshtml:12]: An id here is not required, it is just useful if you want to target these with javascript.

And for user shapes, there isn't an easy id, but we don't need it.

**Comment by deanmarcussen** [comment]
Fixes https://github.com/OrchardCMS/OrchardCore/issues/7526

**Comment by jptissot** [comment]
I would move this in a seperate assembly as I think the User's module is not a CMS module ? It's classified under the OrchardCore.Modules in the VS solution. I don't know enough about the dependency tree, just going with the VS solution's structure.

**Comment by agriffard** [comment]
Is it possible to set those fields in a Register and an Edit Profile page?

**Comment by agriffard** [comment]
Some people could also want to use them on the front instead of the username.
You explained in the doc in Liquid. 
Can you please provide a Razor sample?

Could it be possible to redefine an admin template to override the displayed user name in admin? 
Ex: Display Avatar and First name if set.

**Comment by sebastienros** [comment]
How do permissions work? Can a user access its own profile? Or just part of it?

What about having a feature that would have a custom controller/view to render a user profile on the front-end, with the goal that devs could customize this page and not worry about security, or loading the user. Obviously I'd prefer it to be totally decoupled but that could be a separate guide ;)

**Comment by deanmarcussen** [comment]
Re permissions.

- This issue should probably be closed https://github.com/OrchardCMS/OrchardCore/issues/1025
- The pr that added `IEntity` support for Users, has some good comments about permissions that we could add, that never quite got done I suspect https://github.com/OrchardCMS/OrchardCore/pull/1054

As there isn't permissions other than `ManageUsers` as yet, this feature doesn't have any permission handlers.

As far as guides go... there's room for plenty on users - how to add a custom claim factory (for providing a profile picture via a claim), and then use it in the admin would be a good one @agriffard There's demo code for some of it in the `Demo` project.

And a user profile feature / controller would also work.

Also not far behind is a `UserPickerField`.

**Comment by davidhayden** [comment]
This looks really cool, and I can't wait to try it out!

I couldn't tell from the conversation, but is it possible to make this available on the frontend where a user can manage their own profile?

The UserProfile example in the demo code isn't ideal, because it doesn't work with content editors and requires coding each time a change is made to the profile. This solution, if available on the frontend, would be similar to the Profile Part in O1 that allowed one to simply add fields to a content part that each user was then able to securely access and update on the frontend.

**Comment by sebastienros** [comment]
@davidhayden Right now this PR doesn't have any front end capability. I was suggesting we should add a controller to "display" and only display the profile, such that you can create a custom template file to change it. But no edition since the settings are edited using standard fields which are meant for the admin, and handling front-end edition is too complex in terms of UI. 

My suggestion would be to provide your own form and controller/razor page on the front-end if you need it. At least the current PR provides a way to define the forms on the admin so you can store, see and update custom information for any user.

Or optionally use the forms module and a workflow to let users edit very specific parts of their profile. Custom widgets can also read the profile data of the current logged users. These are just ideas of the possibilities...

**Comment by hishamco** [comment]
From design perspective I'd like to have 3 tabs as following:

1. Content

Which is shown as screenshot above

2. Profile

Which contains: FirstName, LastName, Photo, Email .. etc

3. Others

Which contains other custom user properties something like Mobile, Address, Social media links ... etc. Such properties could be extensible and created programmatically

So far what has been done is great but it needs a little organization as I mentioned above. Another thing which may I create a PR for it, is to create a menu for user profile in the admin theme before logout menu, that allow user to edit his profile

**Comment by deanmarcussen** [comment]
It's great how you can create them all through the admin ui and get complete freedom on how they are designed @hishamco 

Next time I'm preparing a couple of demo screenshots I'll make sure to highlight that they're only an example of what can be done ;)

**Comment by hishamco** [comment]
Great!! waiting for your demos as usual ;)

**Comment by agriffard** [comment]
Can you please add a page to edit your user Profile from the admin, as discussed during the meeting?

**Comment by jptissot** [comment]
> Can you please add a page to edit your user Profile from the admin, as discussed during the meeting?

I think the suggestion was around having a link in the menu to edit your own user.

**Comment by hishamco** [comment]
@agriffard I think what you asking is similar to what I suggested before, I think it should be in another PR I can create a menu item that link to the current user edit page, no need to create another page

**Comment by jptissot** [comment]
> @agriffard I think what you asking is similar to what I suggested before, I think it should be in another PR I can create a menu item that link to the current user edit page, no need to create another page

I think we also need a seperate page. In case the user has access to the admin but does not have permissions to edit users (and roles)

**Comment by hishamco** [comment]
@jptissot I don't think we need a new page for the profile, while the user able to edit his info

> In case the user has access to the admin but does not have permissions to edit users (and roles)

In this we can disable non user profile settings, so the user will not be able to edit his roles

**Comment by deanmarcussen** [comment]
Yes, I will handle the authorization / permissions / display aspects discussed in the meeting as part of this pr.

- separate drivers and views for the current user driver into user editable sections / admin only sections
- changes to user permissions for `EditOwnUser` 
- user information edit controller action for managing the driver
- dynamic permissions for the dynamic sections added via content types

@davidhayden just to clarify what @sebastienros discusses.

- This pr is focusing on the admin, but we will look towards front end display, probably not front end editing (though maybe providing an example or guide), in another pr.
- The custom user settings, while creatable through the admin, are pretty similar to the examples provided in the `OrchardCore.Demo` module, in that for a custom user claim, you'll still need to implement a custom claims factory, to map the user settings data, to a claim.
- So it's mostly about making it easier to create these settings through the admin interface, rather than code, for the moment

File: src/OrchardCore.Modules/OrchardCore.Users/Drivers/UserDisplayDriver.cs [hunk 2]
```diff
@@ -121,46 +128,8 @@ public override async Task<IDisplayResult> UpdateAsync(User user, UpdateEditorCo
...
-                var userWithSameNameId = await _userManager.GetUserIdAsync(userWithSameName);
-                if (userWithSameNameId != model.Id)
-                {
-                    context.Updater.ModelState.AddModelError(string.Empty, S["The user name is already used."]);
-                }
-            }
-
-            if (model.UserName != user.UserName && user.UserName == httpContext.User.Identity.Name)
-            {
-                context.Updater.ModelState.AddModelError(string.Empty, S["Cannot modify user name of the currently logged in user."]);
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Users/Drivers/UserDisplayDriver.cs:146]: @Skrypt I need to understand why you added this, as I have removed it, and forced a new cookie when editing the current user... so allowing the user to change his name.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Users/Drivers/UserDisplayDriver.cs:146]: AFAIK this has been added when the user edit his info (username & role), so it restricts the editing to specific information

**Comment by deanmarcussen** [comment]
From the original pr https://github.com/OrchardCMS/OrchardCore/pull/1054

> Just a little request, could we add new permissions named ManageUserInformation and ManageOwnUserInformation (which is implied by the first one)? When editing a user we then check for ManageUserInformation. The ManageOwnInformation is provided by a permission handler (like for content item Ownership permissions).

So Permissions added, plus dynamic ones for the custom user settings.

Here's some admin profile views

`ManageAllUsers`
Screenshot 2020-11-13 at 16 09 47

`ManageOwnUserInformation` + Manage just that custom setting.
Screenshot 2020-11-13 at 16 10 08

Screenshot 2020-11-13 at 16 10 17

**Comment by davidhayden** [comment]
Thanks for the clarification @sebastienros  and @deanmarcussen 

This looks **awesome** and I look forward to the frontend integration as well.

**Pull request merged**